### PR TITLE
Migrate errata.

### PR DESCRIPTION
**Pull request by goosemania**
closes #6178
https://pulp.plan.io/issues/6178

### CODE REVIEW

File: pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py [hunk 1]
```diff
@@ -189,11 +193,62 @@ class Pulp2Erratum(Pulp2to3Content):
...
+
+        pulp2rpms = Pulp2Rpm.objects.filter(pulp2_id__in=package_pulp2_ids)
+        for pkg in pulp2rpms.iterator():
+            repo_pkg_nevra.append((pkg.name, pkg.epoch or '0', pkg.version, pkg.release, pkg.arch))
+
+        # gather info about available modules
+        modulemd_pulp2_ids = Pulp2RepoContent.objects.filter(
+            pulp2_repository=pulp2_repo,
+            pulp2_content_type_id='modulemd'
+        ).only('pulp2_unit_id').values_list('pulp2_unit_id', flat=True)
```
Comment (dralley) [inline pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py:236]: Is there a particular reason why to use values_list() in a couple of places?

Comment (goosemania) [reply pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py:236]: @dralley , just a way to construct a list of values to feed in the another query.
Any alternative suggestions are welcome.

Comment (dralley) [reply pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py:236]: Actually, never mind, I was going to say that you can just pass in the queryset, but I forgot that that only works with the PK, wheras here we actually have "pulp2_unit_id", so this is in fact probably the simplest way to do it.

File: pulp_2to3_migration/app/plugin/rpm/erratum.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    if checksum_type_id:
+        return (checksum_type_id, checksum)
+
+
+async def get_bool(value):
+    """
+    Convert a value from an erratum to a boolean.
+
+    Sometimes the value is None, some times it's an int, and sometimes it's a string with a boolean
+    value, e.g. "True".
```
Comment (dralley) [inline pulp_2to3_migration/app/plugin/rpm/erratum.py:133]: :(

File: pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py [hunk 3]
```diff
@@ -141,44 +156,88 @@ class Pulp2Erratum(Pulp2to3Content):
...
+            return cls.cached_repo_data[pulp2_repo.pk]
+
+        repo_pkg_nevra = []
+        repo_module_nsvca = []
+
+        # gather info about available packages
+        package_pulp2_ids = Pulp2RepoContent.objects.filter(
+            pulp2_repository=pulp2_repo,
+            pulp2_content_type_id='rpm'
+        ).only('pulp2_unit_id').values_list('pulp2_unit_id', flat=True)
```
Comment (ipanova) [inline pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py:215]: can you add iterator() on this sort of queries?

Comment (dralley) [reply pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py:215]: I'm not sure

**Comment by ipanova** [comment]
i tested with https://repos.fedorapeople.org/pulp/pulp/fixtures/rpm-references-updateinfo/repodata/ and https://repos.fedorapeople.org/pulp/pulp/fixtures/rpm-test-modularity/

1. I have created and synced repos in pulp3
2. then i migrated same content from pulp2

When i compared migrated errata and errata directly synced into pulp3 i have noticed following issues
1) relations to references were not created, i.e references were []
2) updated_date was very funny  "updated_date": "2975-12-07 07:00:00", but not on all records.

```
$ http GET :24817/pulp/api/v3/content/rpm/advisories/?id=RHEA-2012:0060
HTTP/1.1 200 OK
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 1612
Content-Type: application/json
Date: Wed, 15 Apr 2020 17:33:28 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
        {
            "artifact": null,
            "description": "Duck_0.8_Erratum description",
            "fromstr": "errata@redhat.com",
            "id": "RHEA-2012:0060",
            "issued_date": "2018-01-29 16:08:09",
            "pkglist": [
                {
                    "name": "coll_name",
                    "packages": [
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "duck-0.8-1.noarch.rpm",
                            "name": "duck",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "0.8"
                        }
                    ],
                    "shortname": ""
                }
            ],
            "pulp_created": "2020-04-15T17:23:39.493528Z",
            "pulp_href": "/pulp/api/v3/content/rpm/advisories/0b843f1c-abd1-4480-ad89-99930a55ede8/",
            "pushcount": "",
            "reboot_suggested": false,
            "references": [],
            "release": "1",
            "rights": "",
            "severity": "",
            "solution": "",
            "status": "stable",
            "summary": "",
            "title": "Duck_0.8_Erratum",
            "type": "enhancement",
            "updated_date": "2018-07-29 06:00:01",
            "version": "1"
        },
        {
            "artifact": null,
            "description": "Duck_0.8_Erratum description",
            "fromstr": "errata@redhat.com",
            "id": "RHEA-2012:0060",
            "issued_date": "2018-01-29 16:08:09",
            "pkglist": [
                {
                    "name": "collection-0",
                    "packages": [
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "duck-0.8-1.noarch.rpm",
                            "name": "duck",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "0.8"
                        }
                    ],
                    "shortname": ""
                }
            ],
            "pulp_created": "2020-04-15T17:24:18.400958Z",
            "pulp_href": "/pulp/api/v3/content/rpm/advisories/755eec90-f0a8-4975-977e-4b41e18fb275/",
            "pushcount": "",
            "reboot_suggested": false,
            "references": [],
            "release": "1",
            "rights": "",
            "severity": "",
            "solution": "",
            "status": "stable",
            "summary": "",
            "title": "Duck_0.8_Erratum",
            "type": "enhancement",
            "updated_date": "2975-12-07 07:00:00",
            "version": "1"
        }
    ]
}

(pulp) [vagrant@pulp2-nightly-pulp3-source-centos7 ~]$ 

```

second feed testing
```
$ http GET ':24817/pulp/api/v3/content/rpm/advisories/?id=RHEA-2012:0055&repository_version=/pulp/api/v3/repositories/rpm/rpm/768defcb-7ff5-4fb6-961d-9a598a7ad27e/versions/1/'
HTTP/1.1 200 OK
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 1214
Content-Type: application/json
Date: Wed, 15 Apr 2020 18:00:48 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "artifact": null,
            "description": "Sea_Erratum",
            "fromstr": "errata@redhat.com",
            "id": "RHEA-2012:0055",
            "issued_date": "2012-01-27 16:08:06",
            "pkglist": [
                {
                    "name": "collection-0",
                    "packages": [
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "walrus-5.21-1.noarch.rpm",
                            "name": "walrus",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "5.21"
                        },
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "penguin-0.9.1-1.noarch.rpm",
                            "name": "penguin",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "0.9.1"
                        },
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "shark-0.1-1.noarch.rpm",
                            "name": "shark",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "0.1"
                        }
                    ],
                    "shortname": ""
                }
            ],
            "pulp_created": "2020-04-15T17:24:18.407264Z",
            "pulp_href": "/pulp/api/v3/content/rpm/advisories/fd05e251-fb3e-4588-bd46-127ee76ed437/",
            "pushcount": "",
            "reboot_suggested": false,
            "references": [],
            "release": "1",
            "rights": "",
            "severity": "",
            "solution": "",
            "status": "stable",
            "summary": "",
            "title": "Sea_Erratum",
            "type": "security",
            "updated_date": "2012-01-27 16:08:06",
            "version": "1"
        }
    ]
}

(pulp) [vagrant@pulp2-nightly-pulp3-source-centos7 ~]$ http GET ':24817/pulp/api/v3/content/rpm/advisories/?id=RHEA-2012:0055&repository_version=/pulp/api/v3/repositories/rpm/rpm/50fabf63-9bc2-4dfc-80ad-bd538ff09e24/versions/1/'
HTTP/1.1 200 OK
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 1518
Content-Type: application/json
Date: Wed, 15 Apr 2020 18:00:56 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "artifact": null,
            "description": "Sea_Erratum",
            "fromstr": "errata@redhat.com",
            "id": "RHEA-2012:0055",
            "issued_date": "2012-01-27 16:08:06",
            "pkglist": [
                {
                    "name": "1",
                    "packages": [
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "shark-0.1-1.noarch.rpm",
                            "name": "shark",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "0.1"
                        },
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "penguin-0.9.1-1.noarch.rpm",
                            "name": "penguin",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "0.9.1"
                        },
                        {
                            "arch": "noarch",
                            "epoch": "0",
                            "filename": "walrus-5.21-1.noarch.rpm",
                            "name": "walrus",
                            "reboot_suggested": false,
                            "release": "1",
                            "src": "http://www.fedoraproject.org",
                            "sum": "",
                            "sum_type": "",
                            "version": "5.21"
                        }
                    ],
                    "shortname": ""
                }
            ],
            "pulp_created": "2020-04-15T17:49:42.196514Z",
            "pulp_href": "/pulp/api/v3/content/rpm/advisories/7ae666db-7c33-4366-9eb5-8ac835ffccaa/",
            "pushcount": "",
            "reboot_suggested": false,
            "references": [
                {
                    "href": "https://bugzilla.redhat.com/show_bug.cgi?id=1364920",
                    "id": "1364920",
                    "title": "links: Unix domain sockets shared between anonymous and non-anonymous instances",
                    "type": "bugzilla"
                },
                {
                    "href": "https://bugzilla.redhat.com/show_bug.cgi?id=1352033",
                    "id": "1352033",
                    "title": "links-2.13 is available",
                    "type": "bugzilla"
                }
            ],
            "release": "1",
            "rights": "",
            "severity": "",
            "solution": "",
            "status": "stable",
            "summary": "",
            "title": "Sea_Erratum",
            "type": "security",
            "updated_date": "2012-01-27 16:08:06",
            "version": "1"
        }
    ]
}

(pulp) [vagrant@pulp2-nightly-pulp3-source-centos7 ~]$ 
```

**Comment by ipanova** [comment]
```
File "/home/vagrant/devel/pulp-2to3-migration/pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py", line 389, in create_pulp3_c
Apr 15 18:41:14 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[16894]: pulp3content, extra_info = await pulp_2to3_detail_content.create_pulp3_content()
Apr 15 18:41:14 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[16894]: File "/home/vagrant/devel/pulp-2to3-migration/pulp_2to3_migration/app/plugin/content.py", line 233, in migrate_to_pulp3
Apr 15 18:41:14 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[16894]: Traceback (most recent call last):
Apr 15 18:41:14 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[16894]: future: <Task finished coro=<ContentMigrationFirstStage.migrate_to_pulp3() done, defined at /home/vagrant/devel/pulp-2to3-migratio
Apr 15 18:41:14 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[16894]: pulp: asyncio:ERROR: Task exception was never retrieved
```

**Comment by goosemania** [comment]
Dates and references issues has been investigated and fixed. Merging

**Comment by dralley** [comment]
@goosemania Could you try uncommenting this and removing the comment? We should be OK to do so now with the pulpcore change to build from source.

https://github.com/pulp/pulp-2to3-migration/blob/master/setup.py#L10

**Pull request merged**